src/QuadForm/Herm/Genus.jl

```diff
@@ -1806,47 +1807,59 @@ function genus_generators(L::HermLat)
     cocycle = Matrix{elem_type(W)}(undef, ll, ll)
     for i in 1:ll
       for j in 1:ll
-        cocycle = zero(W)
+        cocycle[i,j] = zero(W)
       end
     end
     C = collect(Q0)
     ideals = ideal_type(Rabs)[ q00(C[i]) for i in 1:length(C) ]
     for i in 1:ll
-      for j in 1:ll
-        ij = findfirst(isequal(C[i] * C[j]), C)
-        I = ideals[i] * ideals[j] * inv(ideals[ij])
+      for j in 1:i
+        ij = findfirst(isequal(C[i] + C[j]), C)
+        Iabs = ideals[i] * ideals[j] * inv(ideals[ij])
+	      I = EabstoE(Iabs)
         J = I * inv(a(I))
-        ok, x = isprincipal(J)
-        u = f(nnorm(-(ff\FacElem(nf(RR)(norm(x))))))
+	Jabs = EabstoE\J
+        ok, x = isprincipal(Jabs)
+        u = f(nnorm\(-(ff\FacElem(nf(RR)(norm(x))))))
         x = x * u
         @assert norm(x) == 1
-        y = w(V(Int[ valuation(x - 1, PP[i]) >= VD[i] ? F(0) : F(1) for i in 1:length(PP)]))
-        cocycle[i, j] = y
+	if evaluate(x) == 1
+	  y = w(V(zeros(F,length(PP))))
+	else
+          y = w(V([ valuation(evaluate(x) - 1, PP[i]) >= VD[i] ? F(0) : F(1) for i in 1:length(PP)]))
+        end
+	cocycle[i, j] = y
         cocycle[j, i] = y
       end
     end
 
     S = Tuple{elem_type(Q0), Generic.QuotientModuleElem{elem_type(F)}}[(id(Q0), zero(W))]
-    Work = isdefinite(L) ? typeof(P0)[ P0 ] : typeof(P0)[]
+    Work = def ? typeof(P0)[ P0 ] : typeof(P0)[]
     p = 2
-    while length(S) != order(Q0) * length(W)
+    while length(S) != order(Q0) * dim(W)
       while isempty(Work)
         p = next_prime(p)
         Work = ideal_type(R)[ QQ for QQ in support(p * R) if issplit(QQ) && valuation(bad, QQ) == 0 ]
       end
       P = popfirst!(Work)
       c = q00\P
       i = findfirst(isequal(c), C)
-      I = P * inv(ideals[i])
+      Iabs = P * inv(ideals[i])
+      I = EabstoE(Iabs)
       J = I * inv(a(I))
-      ok, x = isprincipal(J)
-      u = f(nnorm(-(ff\FacElem(nf(RR)(norm(x))))))
+      Jabs = EabstoE\J
+      ok, x = isprincipal(Jabs)
+      u = f(nnorm\(-(ff\FacElem(nf(RR)(norm(x))))))
       x = x * u
       @assert norm(x) == 1
-      y = V(Int[ valuation(x - 1, PP[i]) >= VD[i] ? F(0) : F(1) for i in 1:length(PP)])
+      if evaluate(x) == 1
+        y = w(V(zeros(F,length(PP))))
+      else
+        y = w(V([ valuation(evaluate(x) - 1, PP[i]) >= VD[i] ? F(0) : F(1) for i in 1:length(PP)]))
+      end
       idx = findfirst(isequal(P), PP)
       if idx !== nothing
-        y = V(elem_type(F)[i == idx ? y[i] : y[i] + 1] for i in 1:dim(V)) #w(V([y[idx] = y[idx] + 1
+        y = V(elem_type(F)[i == idx ? y[i] + 1 : y[i]  for i in 1:dim(V)])
```

Comment:
explain

Reply:
I stick to the code of Markus: add 1 to the element at the index `idx`.